core/network/mux.go

```diff
@@ -11,6 +12,51 @@ import (
 // ErrReset is returned when reading or writing on a reset stream.
 var ErrReset = errors.New("stream reset")
 
+type StreamErrorCode uint32
+
+type StreamError struct {
+	ErrorCode      StreamErrorCode
+	Remote         bool
+	TransportError error
+}
+
+func (s *StreamError) Error() string {
+	side := "local"
+	if s.Remote {
+		side = "remote"
+	}
+	if s.TransportError != nil {
+		return fmt.Sprintf("stream reset (%s): code: %d: transport error: %s", side, s.ErrorCode, s.TransportError)
+	}
+	return fmt.Sprintf("stream reset (%s): code: %d", side, s.ErrorCode)
+}
+
+func (s *StreamError) Is(target error) bool {
+	if target == ErrReset {
+		return true
+	}
```

Comment:
Following the discussion
[here](https://github.com/libp2p/go-libp2p/issues/3156#issuecomment-2614215865).
It might make more sense to implement an `Unwrap() []error` method instead.

```suggestion
func (s *StreamError) Unwrap() []error {
	if s.TransportError == nil {
		return []error{ErrReset}
	}
	return []error{ErrReset, s.TransportError}
}
```

Reply:
That makes sense. Done.